src/sources/amqp.rs

```diff
@@ -0,0 +1,565 @@
+    )))
+}
+
+struct Keys<'a> {
+    routing_key: &'a str,
+    routing: &'a str,
+    exchange_key: &'a str,
+    exchange: &'a str,
+    offset_key: &'a str,
+    delivery_tag: i64,
+}
+
+/// Populates the decoded event with extra metadata.
+fn populate_event(
+    event: &mut Event,
+    timestamp: chrono::DateTime<Utc>,
+    keys: &Keys<'_>,
+    log_namespace: LogNamespace,
+) {
+    let log = event.as_mut_log();
+
+    log_namespace.insert_vector_metadata(
+        log,
+        log_schema().timestamp_key(),
+        "ingest_timestamp",
+        timestamp,
+    );
+
+    log_namespace.insert_vector_metadata(
+        log,
+        log_schema().source_type_key(),
+        "source_type",
+        "amqp",
+    );
+
+    log_namespace.insert_source_metadata(
+        "amqp",
+        log,
+        keys.routing_key,
+        "routing",
+        keys.routing.to_string(),
+    );
+
+    log_namespace.insert_source_metadata(
+        "amqp",
+        log,
+        keys.exchange_key,
+        "exchange",
+        keys.exchange.to_string(),
+    );
+
+    log_namespace.insert_source_metadata("amqp", log, keys.offset_key, "offset", keys.delivery_tag);
+}
+
+/// Runs the AMQP source involving the main loop pulling data from the server.
+async fn run_amqp_source(
+    config: AMQPSourceConfig,
+    shutdown: ShutdownSignal,
+    mut out: SourceSender,
+    channel: Channel,
+    log_namespace: LogNamespace,
+    acknowledgements: bool,
+) -> Result<(), ()> {
+    let (finalizer, mut ack_stream) =
+        UnorderedFinalizer::<FinalizerEntry>::maybe_new(acknowledgements, shutdown.clone());
+
+    debug!("Starting amqp source, listening to queue {}.", config.queue);
+    let mut consumer = channel
+        .basic_consume(
+            &config.queue,
+            &config.consumer,
+            lapin::options::BasicConsumeOptions::default(),
+            lapin::types::FieldTable::default(),
+        )
+        .await
+        .map_err(|error| {
+            error!(message = "Failed to consume.", error = ?error, internal_log_rate_secs = 10);
+        })?
+        .fuse();
+    let mut shutdown = shutdown.fuse();
+    loop {
+        tokio::select! {
+            _ = &mut shutdown => break,
+            entry = ack_stream.next() => {
+                if let Some((status, entry)) = entry {
+                    handle_ack(status, entry).await;
+                }
+            },
+            opt_m = consumer.next() => {
+                if let Some(try_m) = opt_m {
+                    match try_m {
+                        Err(error) => {
+                            emit!(AMQPEventError { error });
+                            return Err(());
+                        }
+                        Ok(msg) => {
+                            emit!(AMQPEventReceived {
+                                byte_size: msg.data.len()
+                            });
+
+                            if msg.data.is_empty() {
+                                return Err(());
```

Comment:
Should we fire an internal event for "Error" per https://github.com/vectordotdev/vector/blob/master/docs/specs/instrumentation.md#error  here?

Reply:
I've removed that altogether. I think it should be perfectly valid to receive an empty message.